kinto/plugins/accounts/views.py

```diff
@@ -103,16 +130,51 @@ def process_object(self, new, old=None):
 
         new["password"] = hash_password(new["password"])
 
-        # Administrators can reach other accounts and anonymous have no
-        # selected_userid. So do not try to enforce.
-        if self.context.is_administrator or self.context.is_anonymous:
-            return new
-
         # Do not let accounts be created without usernames.
         if self.model.id_field not in new:
             error_details = {"name": "data.id", "description": "Accounts must have an ID."}
             raise_invalid(self.request, **error_details)
 
+        # If the "account validation" option is enabled, make sure the
+        # `activation-form-url` is set on user creation.
+        # TODO: this might be better suited for a schema. Do we have a way to
+        # dynamically change the schema according to the settings?
+        if self.context.validation and old is None:
+            email_regexp = self.context.validation_email_regexp
+            # Account validation requires that the record ID is an email address.
+            user_email = new[self.model.id_field]
+            if not email_regexp.match(user_email):
+                error_details = {
+                    "name": "data.id",
+                    "description": "Account validation is enabled, and user id must be an email",
+                }
+                raise_invalid(self.request, **error_details)
+
+            if "activation-form-url" not in new:
+                error_details = {
+                    "name": "data.activation-form-url",
+                    "description": "Account validation is enabled, and requires an "
+                    "`activation-form-url` field to be set",
+                }
+                raise_invalid(self.request, **error_details)
+
+            activation_key = str(uuid.uuid4())
+            new["activation-key"] = activation_key
```

Comment:
Haha, I just realized while taking my shower: storing this activation key in the account means the user will get it back as a response to the POST. And will thus be able to POST to the `validate` endpoint without needing the email. Which defeats the whole purpose of validating the email.

So this `activation-key` has to be used straight away while sending the email, but not stored!

Also, I want to be paid overtime.

Reply:
Storing it in the cache backend seems to be a good idea in that case.

Reply:
Or we could generate the activation-key by hashing the record with the server's secret key, which means it could be re-generated when needed (when checking the validation key on the `validate` endpoint, or if the activation link needs to be sent again).

Reply:
> Or we could generate the activation-key by hashing the record with the server's secret key

To make this correctly, the server secret key should be rotqted for each user which means you would still have to store the hash in the cache.

The benefit of using the cache is that you can handle expiration seamlessly.

Reply:
What i see from the code in `kinto/plugins/accounts/authentication.py` is that we're using the server secret to get the `cache_key`. Could we use the exact same technique with the user record (and not only the username) to generate the activation key?

I don't understand why we would have to rotate the secret key? Is it because the cache key is not accessible to the end user, but the activation key is, and would thus allow attackers to get enough activation keys to infer the secret key?

Reply:
I've implemented your suggestion @Natim storing the activation key in the cache, let me know what you think ;)

https://github.com/Kinto/kinto/pull/1982/commits/9dd642ebb038497215dbc0dce7f499aae604d14a

Reply:
Yep seems good.